File: packages/klarna-network-payment/android/src/main/java/com/klarna/mobile/sdk/reactnative/klarnanetwork/payment/KlarnaNetworkPaymentModuleImpl.kt
package com.klarna.mobile.sdk.reactnative.klarnanetwork.payment

import android.os.Handler
import android.os.Looper
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.klarna.mobile.sdk.klarna.network.core.api.common.KlarnaResult
import com.klarna.mobile.sdk.klarna.network.payment.api.presentation.models.KlarnaPaymentPresentationContent
import com.klarna.mobile.sdk.klarna.network.payment.api.payment
import com.klarna.mobile.sdk.reactnative.klarnanetwork.core.KlarnaNetworkCoreModuleImpl
import java.util.concurrent.ConcurrentHashMap

class KlarnaNetworkPaymentModuleImpl(
    private val reactContext: ReactApplicationContext
) {
    companion object {
        const val NAME = "KlarnaNetworkPayment"
        private const val ERROR_ACTIVITY_NOT_FOUND = "Main Activity not found."
        private const val ERROR_INSTANCE_NOT_FOUND = "No instance found for the given instanceId. Call initialize first."
        private const val ERROR_NO_PRESENTATION_CONTENT = "No presentation content found. Call presentationFetch first."
        val storedPresentationContent: ConcurrentHashMap<String, KlarnaPaymentPresentationContent> = ConcurrentHashMap()
    }

    fun initiateWithId(instanceId: String, paymentRequestId: String, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        val activity = reactContext.currentActivity ?: run {
            promise?.reject(NAME, ERROR_ACTIVITY_NOT_FOUND)
            return
        }
        // PaymentControllerImpl calls fragmentManager.setFragmentResultListener() synchronously
        // inside initiate(), which calls lifecycle.addObserver() — must run on the main thread.
        Handler(Looper.getMainLooper()).post {
            sdk.payment.initiate(activity = activity, paymentRequestId = paymentRequestId) { result ->
                when (result) {
                    is KlarnaResult.Success -> promise?.resolve(result.value.toWritableMap())
                    is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
                }
            }
        }
    }

    fun initiateWithData(instanceId: String, data: ReadableMap?, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        val activity = reactContext.currentActivity ?: run {
            promise?.reject(NAME, ERROR_ACTIVITY_NOT_FOUND)
            return
        }
        val paymentRequestData = try {
            data?.let { KlarnaPaymentRequestMapper.buildPaymentRequestData(it) }
        } catch (e: Exception) {
            promise?.reject(NAME, e.message ?: "Invalid payment request data")
            return
        } ?: run {
            promise?.reject(NAME, "data is required")
            return
        }
        // Same as initiateWithId — must run on the main thread for the same reason.
        Handler(Looper.getMainLooper()).post {
            sdk.payment.initiate(activity = activity, paymentRequestData = paymentRequestData) { result ->
                when (result) {
                    is KlarnaResult.Success -> promise?.resolve(result.value.toWritableMap())
                    is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
                }
            }
        }
    }

    // No Handler wrap needed: fetch and cancel dispatch to Dispatchers.Main internally.
    fun fetch(instanceId: String, paymentRequestId: String, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        sdk.payment.fetch(paymentRequestId = paymentRequestId) { result ->
            when (result) {
                is KlarnaResult.Success -> promise?.resolve(result.value.toWritableMap())
                is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    fun cancel(instanceId: String, paymentRequestId: String, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        sdk.payment.cancel(paymentRequestId = paymentRequestId) { result ->
            when (result) {
                is KlarnaResult.Success -> promise?.resolve(result.value.toWritableMap())
                is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    // No Handler wrap needed: fetch dispatches to Dispatchers.Main internally.
    fun presentationFetch(instanceId: String, data: ReadableMap?, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        val presentationData = try {
            data?.let { KlarnaPaymentRequestMapper.buildPresentationData(it) } ?: run {
                promise?.reject(NAME, "data is required")
                return
            }
        } catch (e: Exception) {
            promise?.reject(NAME, e.message ?: "Invalid presentation data")
            return
        }
        sdk.payment.presentation.fetch(data = presentationData) { result ->
            when (result) {
                is KlarnaResult.Success -> {
                    if (KlarnaNetworkCoreModuleImpl.getInstance(instanceId) != null) {
                        storedPresentationContent[instanceId] = result.value
                    }
                    promise?.resolve(result.value.toWritableMap())
                }
                is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    fun presentationHandleLink(instanceId: String, url: String, promise: Promise?) {
        val sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId) ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        val activity = reactContext.currentActivity ?: run {
            promise?.reject(NAME, ERROR_ACTIVITY_NOT_FOUND)
            return
        }
        val content = storedPresentationContent[instanceId] ?: run {
            promise?.reject(NAME, ERROR_NO_PRESENTATION_CONTENT)
            return
        }
        // handleLink opens a browser/web flow and must run on the main thread.
        Handler(Looper.getMainLooper()).post {
            sdk.payment.presentation.handleLink(activity = activity, content = content, url = url) { result ->
                when (result) {
                    is KlarnaResult.Success -> {
                        if (KlarnaNetworkCoreModuleImpl.getInstance(instanceId) != null) {
                            storedPresentationContent[instanceId] = result.value
                        }
                        promise?.resolve(result.value.toWritableMap())
                    }
                    is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
                }
            }
        }
    }

    fun dispose(instanceId: String, promise: Promise?) {
        storedPresentationContent.remove(instanceId)
        promise?.resolve(null)
    }
}

File: packages/klarna-network-payment/ios/Sources/KlarnaNetworkPaymentModuleImpl.swift
import Foundation
import KlarnaNetworkPayment
@_spi(RNKlarnaNetworkCore) import react_native_klarna_network_core

// MARK: - PresentationContentStore

// Written and read from the JS thread and SDK callbacks (arbitrary threads).
// A serial queue serializes all access to prevent data races.
final class PresentationContentStore {
    private let queue = DispatchQueue(label: "com.klarna.rnnetworkpayment.presentationcontent")
    private var storage: [String: KlarnaPaymentPresentationContent] = [:]

    subscript(key: String) -> KlarnaPaymentPresentationContent? {
        get { queue.sync { storage[key] } }
        set { queue.sync { storage[key] = newValue } }
    }

    @discardableResult
    func removeValue(forKey key: String) -> KlarnaPaymentPresentationContent? {
        queue.sync { storage.removeValue(forKey: key) }
    }
}

// MARK: - KlarnaNetworkPaymentModuleImpl

@objc(KlarnaNetworkPaymentModuleImpl)
public class KlarnaNetworkPaymentModuleImpl: NSObject {

    private static let moduleName = "KlarnaNetworkPayment"
    private static let errorInstanceNotFound = "No instance found for the given instanceId. Call initialize first."
    private static let errorInvalidData = "Invalid payment request data."
    private static let errorNoPresentationContent = "No presentation content found. Call presentationFetch first."
    private static let errorAmountRequired = "Amount is required"
    private static let errorCurrencyRequired = "Currency is required"

    private static let storedPresentationContent = PresentationContentStore()

    @objc public func initiateWithId(
        instanceId: String,
        paymentRequestId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        sdk.payment.initiate(paymentRequestId: paymentRequestId) { result in
            switch result {
            case .success(let paymentRequest):
                resolve(paymentRequest.toDictionary())
            case .failure(let error):
                reject(error.name, error.message, nil)
            }
        }
    }

    @objc public func initiateWithData(
        instanceId: String,
        data: NSDictionary,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        guard data["amount"] is NSNumber else {
            reject(Self.moduleName, Self.errorAmountRequired, nil)
            return
        }
        guard let currency = data["currency"] as? String, !currency.isEmpty else {
            reject(Self.moduleName, Self.errorCurrencyRequired, nil)
            return
        }
        guard let paymentRequestData = klarnaPaymentRequestData(from: data) else {
            reject(Self.moduleName, Self.errorInvalidData, nil)
            return
        }
        sdk.payment.initiate(paymentRequestData: paymentRequestData) { result in
            switch result {
            case .success(let paymentRequest):
                resolve(paymentRequest.toDictionary())
            case .failure(let error):
                reject(error.name, error.message, nil)
            }
        }
    }

    @objc public func fetch(
        instanceId: String,
        paymentRequestId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        sdk.payment.fetch(paymentRequestId: paymentRequestId) { result in
            switch result {
            case .success(let paymentRequest):
                resolve(paymentRequest.toDictionary())
            case .failure(let error):
                reject(error.name, error.message, nil)
            }
        }
    }

    @objc public func cancel(
        instanceId: String,
        paymentRequestId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        sdk.payment.cancel(paymentRequestId: paymentRequestId) { result in
            switch result {
            case .success(let paymentRequest):
                resolve(paymentRequest.toDictionary())
            case .failure(let error):
                reject(error.name, error.message, nil)
            }
        }
    }

    @objc public func presentationFetch(
        instanceId: String,
        data: NSDictionary,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        guard let presentationData = klarnaPaymentPresentationData(from: data) else {
            reject(Self.moduleName, Self.errorInvalidData, nil)
            return
        }
        DispatchQueue.main.async {
            sdk.payment.presentation.fetch(data: presentationData) { result in
                switch result {
                case .success(let content):
                    if KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) != nil {
                        Self.storedPresentationContent[instanceId] = content
                    }
                    resolve(content.toDictionary())
                case .failure(let error):
                    reject(error.name, error.message, nil)
                }
            }
        }
    }

    @objc public func dispose(
        instanceId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        Self.storedPresentationContent.removeValue(forKey: instanceId)
        resolve(nil)
    }

    @objc public func presentationHandleLink(
        instanceId: String,
        url: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let sdk = KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        guard let content = Self.storedPresentationContent[instanceId] else {
            reject(Self.moduleName, Self.errorNoPresentationContent, nil)
            return
        }
        DispatchQueue.main.async {
            sdk.payment.presentation.handleLink(content: content, url: url) { result in
                switch result {
                case .success(let newContent):
                    if KlarnaNetworkCoreModuleImpl.getInstance(instanceId: instanceId) != nil {
                        Self.storedPresentationContent[instanceId] = newContent
                    }
                    resolve(newContent.toDictionary())
                case .failure(let error):
                    reject(error.name, error.message, nil)
                }
            }
        }
    }
}
